trackstop.lua

```diff
@@ -0,0 +1,259 @@
+  [MEDIUM] = 20000,
+  [HIGH] = 30000,
+  [HIGHER] = 40000,
+  [MAX] = 50000,
+}
+
+local SPEED_MAP_REVERSE = {}
+for k, v in pairs(SPEED_MAP) do
+  SPEED_MAP_REVERSE[v] = k
+end
+
+local DIRECTION_MAP = {
+  [NORTH] = df.screw_pump_direction.FromSouth,
+  [EAST] = df.screw_pump_direction.FromWest,
+  [SOUTH] = df.screw_pump_direction.FromNorth,
+  [WEST] = df.screw_pump_direction.FromEast,
+}
+
+local DIRECTION_MAP_REVERSE = {}
+for k, v in pairs(DIRECTION_MAP) do
+  DIRECTION_MAP_REVERSE[v] = k
+end
+
+TrackStopOverlay = defclass(TrackStopOverlay, overlay.OverlayWidget)
+TrackStopOverlay.ATTRS{
+  default_pos={x=-71, y=29},
+  default_enabled=true,
+  viewscreens='dwarfmode/ViewSheets/BUILDING/Trap',
+  frame={w=27, h=4},
+  frame_style=gui.MEDIUM_FRAME,
+  frame_background=gui.CLEAR_PEN,
+}
+
+function TrackStopOverlay:getFriction()
+  return dfhack.gui.getSelectedBuilding().friction
+end
+
+function TrackStopOverlay:setFriction(friction)
+  local building = dfhack.gui.getSelectedBuilding()
+
+  building.friction = FRICTION_MAP[friction]
+end
+
+function TrackStopOverlay:getDumpDirection()
+  local building = dfhack.gui.getSelectedBuilding()
+  local use_dump = building.use_dump
+  local dump_x_shift = building.dump_x_shift
+  local dump_y_shift = building.dump_y_shift
+
+  if use_dump == 0 then
+    return NONE
+  else
+    if dump_x_shift == 0 and dump_y_shift == -1 then
+      return NORTH
+    elseif dump_x_shift == 1 and dump_y_shift == 0 then
+      return EAST
+    elseif dump_x_shift == 0 and dump_y_shift == 1 then
+      return SOUTH
+    elseif dump_x_shift == -1 and dump_y_shift == 0 then
+      return WEST
+    end
+  end
+end
+
+function TrackStopOverlay:setDumpDirection(direction)
+  local building = dfhack.gui.getSelectedBuilding()
+
+  if direction == NONE then
+    building.use_dump = 0
+    building.dump_x_shift = 0
+    building.dump_y_shift = 0
+  elseif direction == NORTH then
+    building.use_dump = 1
+    building.dump_x_shift = 0
+    building.dump_y_shift = -1
+  elseif direction == EAST then
+    building.use_dump = 1
+    building.dump_x_shift = 1
+    building.dump_y_shift = 0
+  elseif direction == SOUTH then
+    building.use_dump = 1
+    building.dump_x_shift = 0
+    building.dump_y_shift = 1
+  elseif direction == WEST then
+    building.use_dump = 1
+    building.dump_x_shift = -1
+    building.dump_y_shift = 0
+  end
+end
+
+function TrackStopOverlay:render(dc)
+  if not self:shouldRender() then
+    return
+  end
+
+  local building = dfhack.gui.getSelectedBuilding()
+  local friction = building.friction
+  local friction_cycle = self.subviews.friction
+
+  friction_cycle:setOption(FRICTION_MAP_REVERSE[friction])
+
+  self.subviews.dump_direction:setOption(self:getDumpDirection())
+
+  TrackStopOverlay.super.render(self, dc)
+end
+
+function TrackStopOverlay:shouldRender()
+  local building = dfhack.gui.getSelectedBuilding()
+  return building and building.trap_type == df.trap_type.TrackStop
```

Comment:
if you expect building can be nil, then be sure to call `getSelectedBuilding(true)` instead of just `getSelectedBuilding()` so you don't spam the console with errors

Reply:
on the other hand, though, why would this ever return false? You're already attached to `viewscreens='dwarfmode/ViewSheets/BUILDING/Trap'`

Reply:
Yeah I assume the building will always be there

Reply:
so maybe this function can be removed entirely (and onInput() as well)

Reply:
Hmm, so there is currently a spot where the console can get spammed with errors with the way I have this written.

If you have gm-editor open and try to edit a field that opens a dialog to input the new value, then `getSelectedBuilding()` is returning `nil` and the console is spammed with errors. Maybe I should use `getSelectedBuilding(true)` and guard against it being `nil` after all?

Reply:
maybe that's a good idea, at least for the time being. once the dialogs are updated to use ZScreens, this will no longer be an issue, but I don't know when I'm going to get around to doing that.